### PR TITLE
Bratseth/instances in deployment xml rebased

### PR DESCRIPTION
**Pull request by bratseth**

### CODE REVIEW

**Comment by bratseth** [comment]
The deprecated methods in DeploymentSpec with remove = November is an index into the work that must be done to orchestrate multiple instances.

File: config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java [hunk 1]
```diff
@@ -46,218 +45,197 @@
...
+    /** Returns the instance step containing the given instance name, or null if not present */
+    public DeploymentInstanceSpec instance(String name) {
+        return instance(InstanceName.from(name));
+    }
+
+    /** Returns the instance step containing the given instance name, or null if not present */
+    public DeploymentInstanceSpec instance(InstanceName name) {
+        for (Step step : steps) {
+            if ( ! (step instanceof DeploymentInstanceSpec)) continue;
+            DeploymentInstanceSpec instanceStep = (DeploymentInstanceSpec)step;
```
Comment (jonmv) [inline config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java]:
```suggestion
```

File: config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java [hunk 2]
```diff
@@ -46,218 +45,197 @@
...
+
+    /** Returns the instance step containing the given instance name, or null if not present */
+    public DeploymentInstanceSpec instance(String name) {
+        return instance(InstanceName.from(name));
+    }
+
+    /** Returns the instance step containing the given instance name, or null if not present */
+    public DeploymentInstanceSpec instance(InstanceName name) {
+        for (Step step : steps) {
+            if ( ! (step instanceof DeploymentInstanceSpec)) continue;
```
Comment (jonmv) [inline config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java]:
```suggestion
```

File: config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java [hunk 3]
```diff
@@ -46,218 +45,197 @@
...
+    }
+
+    /** Returns the instance step containing the given instance name, or null if not present */
+    public DeploymentInstanceSpec instance(String name) {
+        return instance(InstanceName.from(name));
+    }
+
+    /** Returns the instance step containing the given instance name, or null if not present */
+    public DeploymentInstanceSpec instance(InstanceName name) {
+        for (Step step : steps) {
```
Comment (jonmv) [inline config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java]:
```suggestion
        for (Step instanceStep : steps) {
```

File: config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java (hunk 1 above)
Comment (jonmv) [reply config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java]: Why doesn't the suggestion support multi-line comments!?

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java [hunk 4]
```diff
@@ -73,67 +84,134 @@ public DeploymentSpec read(Reader reader) {
...
+        Notifications notifications = readNotifications(instanceTag, parentTag);
+
+        // Values where there is no default
+        List<Step> steps = new ArrayList<>();
+        for (Element instanceChild : XML.getChildren(instanceTag))
+            steps.addAll(readNonInstanceSteps(instanceChild, globalServiceId, instanceChild));
+        List<Endpoint> endpoints = readEndpoints(instanceTag);
+
+        // Build and return instances with these values
+        return Arrays.stream(instanceNameString.split("'"))
```
Comment (jonmv) [inline config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java]: Should be split by commas, no?

Comment (jonmv) [inline config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java]:
```suggestion
        return Arrays.stream(instanceNameString.split(","))
```

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiTest.java [hunk 5]
```diff
@@ -771,6 +786,7 @@ public void multiple_endpoints() {
         tester.computeVersionStatus();
         createAthenzDomainWithAdmin(ATHENZ_TENANT_DOMAIN, USER_ID);
         ApplicationPackage applicationPackage = new ApplicationPackageBuilder()
+                .instances("instance1")
```
Comment (jonmv) [inline controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiTest.java]: Should have one with more instances, but I guess we need instance orchestration first :)

**Review by jonmv — COMMENTED** [review]
Forgot the [`.rnc`](https://github.com/vespa-engine/vespa/blob/master/config-model/src/main/resources/schema/deployment.rnc)?

**Comment by jonmv** [comment]
[deployment.rnc](https://github.com/vespa-engine/vespa/blob/master/config-model/src/main/resources/schema/deployment.rnc) and split instance names in spec on `,`. 

Otherwise, LGTM :)

Yes, users of the spec must be updated.

File: config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java (hunk 1 above)
Comment (bratseth) [reply config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java]: Pretty sure they do, anyway now I don't know what your suggestion was  :-)

File: config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java (hunk 3 above)
Comment (bratseth) [reply config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentSpec.java]: Ah, ok I got it. Thanks!

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java (hunk 4 above)
Comment (bratseth) [reply config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java]: Excellent catch, thanks! And this shows a test is missing ..

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiTest.java (hunk 5 above)
Comment (bratseth) [reply controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiTest.java]: There's both a default and instance1 in there somewhere, but yes we need more tests when we do the orchestration part at least (but perhaps not in the API).

**Comment by bratseth** [comment]
I'll create a separate PR for the schema update, this one is big enough and should not cause anything to change.

**Comment by jonmv** [comment]
Great, well done! :)

**Comment by freva** [comment]
cfg1 in cd-aws-us-east-1a is currently failing with:
```
[2019-10-07 14:53:28.564] WARNING : configserver     Container.com.yahoo.vespa.config.server.modelfactory.ModelsBuilder 
        Input error when building model 
        exception=
        java.lang.IllegalArgumentException: No instance 'default-t' in deployment.xml'. Instances: default
        at com.yahoo.config.application.api.DeploymentSpec.requireInstance(DeploymentSpec.java:228)
        at com.yahoo.vespa.model.container.xml.ContainerModelBuilder.lambda$addIdentityProvider$11(ContainerModelBuilder.java:896)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at com.yahoo.vespa.model.container.xml.ContainerModelBuilder.addIdentityProvider(ContainerModelBuilder.java:895)
        at com.yahoo.vespa.model.container.xml.ContainerModelBuilder.lambda$addAthensCopperArgos$0(ContainerModelBuilder.java:218)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at com.yahoo.vespa.model.container.xml.ContainerModelBuilder.addAthensCopperArgos(ContainerModelBuilder.java:217)
        at com.yahoo.vespa.model.container.xml.ContainerModelBuilder.addClusterContent(ContainerModelBuilder.java:200)
        at com.yahoo.vespa.model.container.xml.ContainerModelBuilder.doBuild(ContainerModelBuilder.java:146)
        at com.yahoo.vespa.model.container.xml.ContainerModelBuilder.doBuild(ContainerModelBuilder.java:96)
        at com.yahoo.config.model.builder.xml.ConfigModelBuilder.build(ConfigModelBuilder.java:74)
        at com.yahoo.config.model.ConfigModelRepo.buildModel(ConfigModelRepo.java:207)
        at com.yahoo.config.model.ConfigModelRepo.buildModels(ConfigModelRepo.java:193)
        at com.yahoo.config.model.ConfigModelRepo.readConfigModels(ConfigModelRepo.java:155)
        at com.yahoo.config.model.ConfigModelRepo.readConfigModels(ConfigModelRepo.java:81)
        at com.yahoo.vespa.model.VespaModel.<init>(VespaModel.java:177)
        at com.yahoo.vespa.model.VespaModel.<init>(VespaModel.java:152)
        at com.yahoo.vespa.model.VespaModelFactory.buildModel(VespaModelFactory.java:127)
        at com.yahoo.vespa.model.VespaModelFactory.createModel(VespaModelFactory.java:95)
        at com.yahoo.vespa.config.server.modelfactory.ActivatedModelsBuilder.buildModelVersion(ActivatedModelsBuilder.java:106)
        at com.yahoo.vespa.config.server.modelfactory.ActivatedModelsBuilder.buildModelVersion(ActivatedModelsBuilder.java:47)
        at com.yahoo.vespa.config.server.modelfactory.ModelsBuilder.buildModelVersions(ModelsBuilder.java:160)
        at com.yahoo.vespa.config.server.modelfactory.ModelsBuilder.buildModels(ModelsBuilder.java:106)
        at com.yahoo.vespa.config.server.session.RemoteSession.loadApplication(RemoteSession.java:65)
        at com.yahoo.vespa.config.server.session.RemoteSession.ensureApplicationLoaded(RemoteSession.java:73)
        at com.yahoo.vespa.config.server.session.RemoteSessionRepo.loadSessionIfActive(RemoteSessionRepo.java:168)
        at com.yahoo.vespa.config.server.session.RemoteSessionRepo.sessionAdded(RemoteSessionRepo.java:147)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
        at com.yahoo.vespa.config.server.session.RemoteSessionRepo.initializeSessions(RemoteSessionRepo.java:114)
        at com.yahoo.vespa.config.server.session.RemoteSessionRepo.<init>(RemoteSessionRepo.java:74)
        at com.yahoo.vespa.config.server.tenant.TenantBuilder.createRemoteSessionRepo(TenantBuilder.java:132)
        at com.yahoo.vespa.config.server.tenant.TenantBuilder.build(TenantBuilder.java:72)
        at com.yahoo.vespa.config.server.tenant.TenantRepository.createTenant(TenantRepository.java:207)
        at com.yahoo.vespa.config.server.tenant.TenantRepository.createTenant(TenantRepository.java:198)
        at com.yahoo.vespa.config.server.tenant.TenantRepository.lambda$bootstrapTenants$2(TenantRepository.java:169)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by bratseth** [comment]
Thanks. Looks like we need some special handling for tester applications. Reverting for now.

**Comment by jonmv** [comment]
Right, didn't think of this. 

Perhaps we should just update the generated deployment spec for the tester to _be_ the `default-t` instance?

**Comment by bratseth** [comment]
Yes, where is this?

**Comment by jonmv** [comment]
I can make a separate PR. 

However, there are other cases where the application ID instance doesn't match the deployment spec:

1. Unorchestrated deployments, i.e., deployments to `dev`/`perf`.
1. Any already existing deployments under a non-`"default"` instance.

The latter we can probably break, but the former we have an issue with, since the Athenz services are defined in `deployment.xml` (rather than `services.xml`).

**Comment by jonmv** [comment]
Perhaps we need a set of default settings for the few things that make sense outside production deployments, like `major-version` and `athenz-domain` and `-service`? When an instance isn't found by the container builder, it uses the default instead. This would also solve the tester-case.

**Comment by bratseth** [comment]
Those three can be set on the deployment.xml root. If that's all the tester need I just need to change to lookup without instance in those cases so you don't need to do anything.

And yes, manual deployments shouldn't need something else.

**Comment by jonmv** [comment]
Go ahead :)

**Pull request merged**